File: LeetCode/Medium/0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> pq = new PriorityQueue<>(k, Comparator.reverseOrder());

        Arrays.stream(points)
                .map(Point::new)
                .forEach(point -> {
                    if (pq.size() == k && point.compareTo(pq.peek()) < 0) {
                        pq.poll();
                    }

                    if (pq.size() < k) {
                        pq.offer(point);
                    }
                });

        return pq.stream()
                .map(point -> point.point)
                .toArray(int[][]::new);
    }

    static class Point implements Comparable<Point> {

        int[] point;
        double dist;

        Point(int[] point) {
            this.point = point;
            this.dist = Math.sqrt(Math.pow(point[0], 2) + Math.pow(point[1], 2));
        }

        @Override
        public int compareTo(Point o) {
            return Double.compare(dist, o.dist);
        }

    }

}

File: LeetCode/Medium/0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.kt
import java.util.*
import kotlin.math.pow
import kotlin.math.sqrt

class Solution {

    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val pq = PriorityQueue<IntArray>(k, compareByDescending { distance(it) })

        for (point in points) {
            if (pq.size < k) {
                pq.offer(point)
                continue
            }

            if (distance(point) < distance(pq.peek())) {
                pq.poll()
                pq.offer(point)
            }
        }

        return pq.toTypedArray()
    }

    fun distance(point: IntArray): Double {
        val x = point[0].toDouble()
        val y = point[1].toDouble()

        return sqrt(x.pow(2) + y.pow(2))
    }
}
